### PR TITLE
[BUG] Fix backfill of custom function in window_feature

### PR DESCRIPTION
**Pull request by toandaominh1997**
#### What does this implement/fix? Explain your changes.

"In this implementation, within the _window_feature function, if the summarizer is a callable function, it will perform a rolling operation and apply the custom function. However, in the backfill stage, which the previous author set up to use 'Z' for backfilling, the custom function is not applied.
#### Does your contribution introduce a new dependency? If yes, which one?
No

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks like a genuine bug, it probably was introduced in #6081 that fixed another bug.

Could you kindly add a test that checks that this bug is present?

**Comment by fkiraly** [comment]
FYI @benHeid, as you also reviewed #6081

**Comment by toandaominh1997** [comment]
@fkiraly @benHeid I have already added a unit test with the output provider. Please consider reviewing this PR again.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks.

If there's only a single value for the variables (`kwargs`, `column_names`, etc), and no complex fixture generation logic, I would prefer if you assign them inside the test.

**Comment by toandaominh1997** [comment]
@fkiraly  You want to remove 

> Thanks.
> 
> If there's only a single value for the variables (`kwargs`, `column_names`, etc), and no complex fixture generation logic, I would prefer if you assign them inside the test.
In the future, if anyone wants to test WindowSummarizer with the output provider, they can use my function for this case. If you have any ideas or feedback, please let me know.

**Review by fkiraly — APPROVED** [review]
Sure, why not.

FYI, the next release is likely in a week or so, until then I'd appreciate if @benHeid and/or @ShreeshaM07 have a look given the relevance of the recent PR.

**Comment by toandaominh1997** [comment]
@fkiraly I have fixed the pre-commit in code-quality. Please help me with approval again. Thanks

File: sktime/transformations/series/tests/test_window_summarizer.py [hunk 1]
```diff
@@ -194,3 +194,30 @@ def test_wrong_column():
     transformer = WindowSummarizer(target_cols=["dummy"])
     Xt = transformer.fit_transform(X_ll_train)
     return Xt
+
+def count_unique(x):
+  return len(np.unique(x))
+kwargs_custom_function = {
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply sktime/transformations/series/tests/test_window_summarizer.py:200]: the precommits should complain about this anyways - looks like they aren't setup correctly, locally?

Guide here: https://www.sktime.net/en/stable/developer_guide/coding_standards.html

[the remark this answers is not in the document]
Comment (toandaominh1997) [reply sktime/transformations/series/tests/test_window_summarizer.py:200]: Sorry=)) My setup uses space = 2. I have modified it. Thanks

**Comment by toandaominh1997** [comment]
@fkiraly It seems like this PR passed all test cases (docs, code quality). I don't know why all of the other CIs were pushed to the queue. Please consider fixing it

**Comment by yarnabrina** [comment]
We run all tests on main branch UTC Sunday 12, and this ensures we are getting higher test coverage. In PR's we run tests only for the modified sections after satisfying some basic criteria (linting, no global imports of soft dependencies, etc.).

It can't be changed right now, and slowness is to be expected. If you can share some CI enhancement ideas, that'll be much appreciated at #5706.

**Comment by ShreeshaM07** [comment]
Apologies for missing out on this bug for the custom `summarizer` function. But I think this PR has not completely solved the issue yet as it gives an error on running `./examples/01c_forecasting_hierarchical_global.ipynb` file . 

```
import numpy as np


def count_gt130(x):
    """Count how many observations lie above threshold 130."""
    return np.sum((x > 700)[::-1])


kwargs = {
    "lag_feature": {
        "lag": [0],
        count_gt130: [[2, 3]],
        "std": [[1, 4]],
    }
}

transformer = WindowSummarizer(**kwargs)
y_transformed = transformer.fit_transform(y_train)

y_transformed.head(10)
```
The error that comes up is 
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[26], line 18
      9 kwargs = {
     10     "lag_feature": {
     11         "lag": [0],
   (...)
     14     }
     15 }
     17 transformer = WindowSummarizer(**kwargs)
---> 18 y_transformed = transformer.fit_transform(y_train)
     20 y_transformed.head(10)

File [~/sktime_shreesha/sktime/sktime/transformations/base.py:681](https://file+.vscode-resource.vscode-cdn.net/home/shreesha/sktime_shreesha/sktime/examples/~/sktime_shreesha/sktime/sktime/transformations/base.py:681), in BaseTransformer.fit_transform(self, X, y)
    616 """Fit to data, then transform it.
    617 
    618 Fits the transformer to X and y and returns a transformed version of X.
   (...)
    677         Example: i-th instance of the output is the i-th window running over `X`
    678 """
    679 # Non-optimized default implementation; override when a better
    680 # method is possible for a given algorithm.
--> 681 return self.fit(X, y).transform(X, y)

File [~/sktime_shreesha/sktime/sktime/transformations/base.py:595](https://file+.vscode-resource.vscode-cdn.net/home/shreesha/sktime_shreesha/sktime/examples/~/sktime_shreesha/sktime/sktime/transformations/base.py:595), in BaseTransformer.transform(self, X, y)
...
   2469         yield recode_for_categories(
   2470             self.codes[i], self.levels[i], new_levels[i], copy=copy
   2471         )

AssertionError: Length of new_levels (3) must be <= self.nlevels (2)
```
Please let me know if it is reproduced in yours as well @toandaominh1997

**Comment by ShreeshaM07** [comment]
So from my debugging so far I think I have found the place where the error is originating. It is in the `_transform` function in the `summarize.py`. The exact line which is not getting executed and producing error instead is 
```
        Xt_return = pd.concat([Xt_out_df, X.drop(target_cols, axis=1)], axis=1)
```

I had encountered the same error in PR #6037 but it was for the pre-defined `summarizer` functions like mean,std. I fixed it when I kept the order for Multi-Indexed DataFrames to be 1st `shift` then `rolling` else for other DataFrames the order is `rolling` first then `shift`.
But I had not implemented this order in the custom `summarizer` function. So I think that is producing the current error. I will look into this.

**Comment by fkiraly** [comment]
hm, indeed. We should find out why it fails the notebook test, and ensure that equivalent coverage is added in `TestAllTransformers` if it is indeed not being caught.

**Comment by toandaominh1997** [comment]
Hi Team,

I have thoroughly reviewed the _window_feature method within our WindowSummarizer class and identified an issue with duplicated indices when using the apply method on pandas.core.groupby.SeriesGroupBy objects. This duplication occurred during the application of custom functions, impacting the integrity of our data indexing.

To address this, I've replaced all instances of the apply method with transform. This change effectively prevents index duplication, maintaining the original indexing structure throughout the operation. The updated approach has been integrated and tested thoroughly; it has passed all our existing test cases without introducing new issues.

FYI: @fkiraly  @ShreeshaM07

**Comment by fkiraly** [comment]
Could you kindly explain under which conditions index duplication occurs, and give an example?

We should also test for this being fixed.

**Comment by toandaominh1997** [comment]
@fkiraly You can check example like this
```python
import numpy as np
import pandas as pd 
from sktime.transformations.series.summarize import WindowSummarizer

y = pd.read_pickle("examples/global_fc/y.pkl")
def count_gt130(x):
    """Count how many observations lie above threshold 130."""
    return np.sum((x > 130)[::-1])


kwargs = {
    "lag_feature": {
        count_gt130: [[1, 3]],
        "lag": [1],
    }
}
window_summarizer = WindowSummarizer(**kwargs)
output = window_summarizer.fit_transform(y)

output1 = y.groupby(level = [0]).rolling(3, min_periods = 3).apply(count_gt130, raw = True)

output2 = y.groupby(level = [0]).transform(lambda x: x.rolling(window = 3, min_periods = 3).apply(count_gt130, raw = True))

print(f"""
output: \n {output}
output1:\n {output1}
output2:\n {output2.tail(100)}
""")
```

Output:
```
output: 
                       y_count_gt130_1_3  y_lag_1
instances timepoints                            
1         2016-03-15                NaN      NaN
          2016-03-16                NaN   756.67
          2016-03-17                NaN   679.13
          2016-03-18                3.0   633.40
          2016-03-19                3.0  1158.04
...                                 ...      ...
2         2016-04-19                3.0  2407.36
          2016-04-20                3.0  2008.23
          2016-04-21                3.0  1614.98
          2016-04-22                3.0  1597.51
          2016-04-23                3.0  1994.69

[80 rows x 2 columns]
output1:
                                   y
instances instances timepoints     
1         1         2016-03-15  NaN
                    2016-03-16  NaN
                    2016-03-17  3.0
                    2016-03-18  3.0
                    2016-03-19  3.0
...                             ...
2         2         2016-04-19  3.0
                    2016-04-20  3.0
                    2016-04-21  3.0
                    2016-04-22  3.0
                    2016-04-23  3.0

[80 rows x 1 columns]
output2:
                         y
instances timepoints     
1         2016-03-15  NaN
          2016-03-16  NaN
          2016-03-17  3.0
          2016-03-18  3.0
          2016-03-19  3.0
...                   ...
2         2016-04-19  3.0
          2016-04-20  3.0
          2016-04-21  3.0
          2016-04-22  3.0
          2016-04-23  3.0

[80 rows x 1 columns]
```

**Comment by ShreeshaM07** [comment]
> Hi Team,
> 
> I have thoroughly reviewed the _window_feature method within our WindowSummarizer class and identified an issue with duplicated indices when using the apply method on pandas.core.groupby.SeriesGroupBy objects. This duplication occurred during the application of custom functions, impacting the integrity of our data indexing.
> 
> To address this, I've replaced all instances of the apply method with transform. This change effectively prevents index duplication, maintaining the original indexing structure throughout the operation. The updated approach has been integrated and tested thoroughly; it has passed all our existing test cases without introducing new issues.
> 
> FYI: @fkiraly @ShreeshaM07

Yes, this change indeed works and passes all existing test cases for Hierarchical and non Hierarchical DataFrames. But would it not be more convenient for future debugging if it is kept in the earlier structure with separate if conditions for Hierarchical and non Hierarchical DataFrames ?

**Comment by toandaominh1997** [comment]
> > Hi Team,
> > I have thoroughly reviewed the _window_feature method within our WindowSummarizer class and identified an issue with duplicated indices when using the apply method on pandas.core.groupby.SeriesGroupBy objects. This duplication occurred during the application of custom functions, impacting the integrity of our data indexing.
> > To address this, I've replaced all instances of the apply method with transform. This change effectively prevents index duplication, maintaining the original indexing structure throughout the operation. The updated approach has been integrated and tested thoroughly; it has passed all our existing test cases without introducing new issues.
> > FYI: @fkiraly @ShreeshaM07
> 
> Yes, this change indeed works and passes all existing test cases for Hierarchical and non Hierarchical DataFrames. But would it not be more convenient for future debugging if it is kept in the earlier structure with separate if conditions for Hierarchical and non Hierarchical DataFrames ?

with transform in pandas, it performs a group-specific computation and returns a DataFrame with the same index as the original DataFrame. So you don't need to care about Hierarchical and non-Hierarchical in _window_feature function
If you have any ideas to improve the codebase. Please let's me know

**Comment by toandaominh1997** [comment]
@fkiraly Please help me review this PR

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks, @toandaominh1997, and thanks @ShreeshaM07 for further review.

Overall, I think this is on track for the release next week, we should just add the "duplicated index" test case as a test, and in its docstring refer to the bug issue.

**Comment by toandaominh1997** [comment]
@fkiraly I have updated with additional more unit test on panel series and implemented to check for duplicate index names. Please help me review this PR again. Thanks

**Review by fkiraly — APPROVED** [review]
Looks good to me, thanks!

Would also appreciate a review by @yarnabrina and @ShreeshaM07 who commented earlier.
